addon/components/au-file-upload.js

```diff
@@ -0,0 +1,147 @@
+
+export default class AuFileUploadComponent extends Component {
+  @service() store;
+  @service() fileQueue;
+  @tracked uploadErrorData = [];
+
+  constructor() {
+    super(...arguments);
+    this.queueName = `${guidFor(this)}-fileUploads`;
+  }
+
+  ////////
+  // Properties which may be overridden
+  ////////
+  get uploadingMsg() {
+    return `Bezig met het opladen van ${this.fileQueue.files.length} bestand(en). (${this.fileQueue.progress}%)`;
+  }
+
+  get title() {
+    return this.args.title || 'Bestanden toevoegen';
+  }
+
+  get helpTextDragDrop() {
+    return (
+      this.args.helpTextDragDrop ||
+      'Sleep de bestanden naar hier om toe te voegen'
+    );
+  }
+
+  get helpTextFileNotSupported() {
+    return (
+      this.args.helpTextFileNotSupported ||
+      'Dit bestandsformaat wordt niet ondersteund.'
+    );
+  }
+
+  get endPoint() {
+    return this.args.endPoint || '/files';
+  }
+
+  get modelName() {
+    return this.args.modelName || 'file';
+  }
+  get maxFileSizeMB() {
+    return this.args.maxFileSizeMB || 20;
+  }
+
+  ////////
+  // Error handling and introspection
+  ////////
+  get hasErrors() {
+    return this.uploadErrorData.length > 0;
+  }
+
+  // uploadFileTask = {};
+
+  @task({ enqueue: true, maxConcurrency: 3 })
+  *uploadFileTask(file) {
+    this.notifyQueueUpdate();
+    try {
+      const response = yield file.upload(this.endPoint, {
+        'Content-Type': 'multipart/form-data',
+      });
+      const uploadedFile = yield this.store.findRecord(
+        this.modelName,
+        response.body.data.id
+      );
+      return uploadedFile;
+    } catch (e) {
+      this.uploadErrorData = [...this.uploadErrorData, { filename: file.name }];
+      this.removeFileFromQueue(file);
+      return null;
+    }
+  }
+
+  hasValidationErrors(file) {
+    // TODO: see if we can split the side-effects from the
+    // non-side-effects.  The name doesn't suggest this will yield a
+    // state change.
+    if (file.size > this.maxFileSizeMB * Math.pow(1024, 2)) {
+      this.uploadErrorData = [
+        ...this.uploadErrorData,
+        {
+          filename: file.name,
+          error: `Bestand is te groot (max ${this.maxFileSizeMB} MB)`,
+        },
+      ];
+      this.removeFileFromQueue(file);
+      return true;
+    }
+    return false;
+  }
+
+  resetErrors() {
+    this.uploadErrorData = [];
+  }
+
+  click() {
+    this.resetErrors();
+  }
+
+  ////////
+  // File queue
+  ////////
+  removeFileFromQueue(file) {
+    let q = this.fileQueue.queues.find((q) => q.name === this.queueName);
+    q.remove(file);
+  }
+
+  calculateQueueInfo() {
+    const filesQueueInfo = {
+      isQueueEmpty: this.uploadFileTask.state === 'idle',
+    };
+    return filesQueueInfo;
+  }
+
+  notifyQueueUpdate() {
+    if (this.args.onQueueUpdate) {
+      this.args.onQueueUpdate(this.calculateQueueInfo());
+    }
+  }
+
+  ////////
+  // Actions
+  ////////
+  @action
```

Comment:
I would convert this to a task as well. I think this version can cause issues if the component is destroyed before it completes (and no "destroyed" check is done after the async code completes).